notes/plutus-core/Bytestrings.md

```diff
@@ -0,0 +1,116 @@
+this would add quite a lot of complexity that we probably don't want.  [However,
+SCP-1091 proposes bit operations for Plutus Core, and we might want some kind of
+`Word` type for that.]
+
+Fortunately Haskell also provides the `Data.ByteString.Char8` interface to its
+`ByteString` type; this provides, for instance, functions `pack :: String ->
+ByteString` and `unpack :: ByteString -> String` which convert between `String`
+and `ByteString`.  We already have support for strings and bytestrings in Plutus
+Core, so this looks like a more promising way to provide extended bytestring
+functionality.
+
+The `pack` function operates on the ASCII codes of the characters in a string, so
+`pack "91ef22"` would give you the bytestring `#393165663232`, not `#91ef22`;
+however you could get the latter via `pack "\x91\xef\x22"` for example.  Also,
+Strings are lists of Unicode characters and `pack` truncates these to the lowest
+8 bits.
+
+The `Char8` interface also provides an `IsString` instance for `ByteString`, and
+this effectively gives you simple concrete syntax: using `OverloadedStrings`,
+`"\x91\xef\x22"` can be automatically converted into a Bytestring.
+
+### What should we add to PLC?
+
+The simplest way to improve bytestring support in PLC would probably be to add a
+`stringToBytestring` builtin implemented in terms of `Data.ByteString.Char8`.
+For efficiency reasons we should probably use built-in PLC strings rather than
+Scott strings.
+
+As noted above, support for built-in strings is limited, so we may also need to
+add some extra builtins for these at both the PlutusTx and PLC levels.
+
+Here's an initial proposal for some new Plutus Core builtins to improve bytestring support.
+
+  * `stringToBytestring`, converting built-in PLC strings to bytestrings and implemented using `pack`.
+     I think that in conjuction with `OverloadedStrings` this would make it reasonably easy to convert
+     "literal" Haskell ByteStrings to PLC bytestrings.
+  * `integerToChar`, implemented using `Data.Char.chr`.  This could be used to build bytestrings at execution time.
+    `consChar` (see below) would help with this.
+  * A `consChar` operation allowing you to prepend a character to a PLC string. This would help with
+    building bytestrings dynamically.  Also, `PlutusTx.String` contains
+    a `stringToBuiltinString` function which converts Haskell `String` to `Builtin.String` by
+    repeated application of `charToString` and `appendString`, which is not very efficient.  I think that
+   `consChar` would allow us to implement this a low more efficiently.
+  * Alternatively, a `consByte` operation directly adding a character or integer to the front of a bytestring.
+    `Data.Bytestring.Char8.cons` would be helpful for this.
+  * Perhaps also
+     * A `lengthOfBytestring` function
+     * An `indexBytestring` function allowing you to retrieve the character or integer at a given position in a bytestring.
```

Comment:
I guess we have to be careful with these functions to make it clear that it's the `Char8` version of length/index.